### PR TITLE
Change button state before event process

### PR DESCRIPTION
**Pull request by hitaishi19**
## Description

Save button does not show loading state when clicking on it. Instead it remains still for sometime and then shows loading state. This is due to event being fully processed before setting the loading class and disabling the button.

## Issues

- https://github.com/silverstripe/silverstripe-admin/issues/1865

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [x] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
I've added the pull request template back into the PR description. Can you please fill it in, including ticking all of the checkboxes that apply? This will help you identify things that need to be done before this is ready to be reviewed (for example [build the dist javascript](https://docs.silverstripe.org/en/contributing/build_tooling/)).

Please also retarget this PR to the `2.3` branch so that it can be included as a patch release rather than waiting for the next minor release.
You may need to [reset your commits](https://gist.github.com/GuySartorelli/74627843aeb75e12b777b1f44cb22d2f) after retargetting the PR.


I have also tested this locally and found two problems:
1. Instead of swapping to the "loading" appearance, the button goes completely blank:
    image
    It only uses the full loading styling after the `beforesubmitform` stuff has finished.
    An easy way to test this is to add a `sleep(5)` into the `validate()` method of an elemental block, then make sure to edit the block before hitting save.
1. If an elemental block has a validation error, the button stays disabled forever.

**Comment by GuySartorelli** [comment]
To fix 1, probably you need to move everything in the `if($(button).is('button')) {` conditional block.

To fix 2, I suspect the `clearButton()` function needs to be called in the `if (!success) {` conditional block just before it returns `false`.

File: client/src/legacy/LeftAndMain.js [hunk 1]
```diff
@@ -499,15 +517,12 @@ $.entwine('ss', function($) {
...
           return false;
         }
 
         self.trigger('submitform', {form: form, button: button});
 
-        // set button to "submitting" state
-        $(button).addClass('btn--loading loading');
-        $(button).prop('disabled', true);
-
         if($(button).is('button')) {
```
Comment (ChloeSartorelli) [inline client/src/legacy/LeftAndMain.js]: This condition and the code in this block is now duplicated - does it _need_ to be in both places, or can this second block be removed?

If it does need to be in both places separately, please abstract it to a function call, and add some in-line code comments explaining _why_ it needs to happen both here _and_ above.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: client/src/legacy/LeftAndMain.js [hunk 2]
```diff
@@ -460,6 +460,24 @@ $.entwine('ss', function($) {
...
+              "</div>"
+          )
+        );
+
+        $(button).css($(button).outerWidth() + "px");
+      }
+
+      // set button to "submitting" state
+      $(button).addClass('btn--loading loading');
+      $(button).prop('disabled', true);
```
Comment (ChloeSartorelli) [inline client/src/legacy/LeftAndMain.js]: This used to happen before the conditional block. That should probably remain the case.

File: client/src/legacy/LeftAndMain.js [hunk 3]
```diff
@@ -460,6 +460,24 @@ $.entwine('ss', function($) {
       // default to first button if none given - simulates browser behaviour
       if(!button) button = this.find('.btn-toolbar :submit:first');
 
+      if ($(button).is("button")) {
```
Comment (ChloeSartorelli) [inline client/src/legacy/LeftAndMain.js]:
```suggestion
      if ($(button).is('button')) {
```

File: client/src/legacy/LeftAndMain.js [hunk 4]
```diff
@@ -460,6 +460,24 @@ $.entwine('ss', function($) {
...
+          $(
+            '<div class="btn__loading-icon">' +
+              '<span class="btn__circle btn__circle--1"></span>' +
+              '<span class="btn__circle btn__circle--2"></span>' +
+              '<span class="btn__circle btn__circle--3"></span>' +
+              "</div>"
+          )
+        );
+
+        $(button).css($(button).outerWidth() + "px");
```
Comment (ChloeSartorelli) [inline client/src/legacy/LeftAndMain.js]:
```suggestion
        $(button).css($(button).outerWidth() + 'px');
```

**Comment by GuySartorelli** [comment]
Problem 2 from https://github.com/silverstripe/silverstripe-admin/pull/1866#pullrequestreview-2483336309 is still happening. Did you test that scenario?

**Comment by hitaishi19** [comment]
> Problem 2 from [#1866 (review)](https://github.com/silverstripe/silverstripe-admin/pull/1866#pullrequestreview-2483336309) is still happening. Did you test that scenario?

This is not reproducing to me, can you please share the reproduction steps?

**Comment by GuySartorelli** [comment]
1. Create a block
1. Add some validation to an elemental block which will fail
    ```php
    public function validate()
    {
        $result = parent::validate();
        $result->addFieldError('Title', 'FAILURE');
        return $result;
    }
    ```
1. Make a change to the block (or else validation won't trigger)
1. Click "save" on the page.

File: client/src/legacy/LeftAndMain.js [hunk 5]
```diff
@@ -460,6 +460,24 @@ $.entwine('ss', function($) {
...
+      $(button).prop('disabled', true);
+
+      if ($(button).is('button')) {
+        $(button).append(
+          $(
+            '<div class="btn__loading-icon">' +
+              '<span class="btn__circle btn__circle--1"></span>' +
+              '<span class="btn__circle btn__circle--2"></span>' +
+              '<span class="btn__circle btn__circle--3"></span>' +
+              "</div>"
```
Comment (ChloeSartorelli) [inline client/src/legacy/LeftAndMain.js]:
```suggestion
              '</div>'
```
Sorry, didn't notice this one before

**Review by GuySartorelli — APPROVED** [review]
LGTM, works well locally. Thanks for taking the time to implement this fix.

**Pull request merged**